sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCast.scala

```diff
@@ -0,0 +1,214 @@
+ * `fromType`, respectively. If this is true then it means we can safely cast `value` to `fromType`
+ * and thus able to move the cast to the literal side.
+ *
+ * If the `value` is not within range `(min, max)`, the rule breaks the scenario into different
+ * cases and try to replace each with simpler constructs.
+ *
+ * if `value > max`, the cases are of following:
+ *  - `cast(exp, ty) > value` ==> if(isnull(exp), null, false)
+ *  - `cast(exp, ty) >= value` ==> if(isnull(exp), null, false)
+ *  - `cast(exp, ty) === value` ==> if(isnull(exp), null, false)
+ *  - `cast(exp, ty) <=> value` ==> false
+ *  - `cast(exp, ty) <= value` ==> if(isnull(exp), null, true)
+ *  - `cast(exp, ty) < value` ==> if(isnull(exp), null, true)
+ *
+ * if `value == max`, the cases are of following:
+ *  - `cast(exp, ty) > value` ==> if(isnull(exp), null, false)
+ *  - `cast(exp, ty) >= value` ==> exp == max
+ *  - `cast(exp, ty) === value` ==> exp == max
+ *  - `cast(exp, ty) <=> value` ==> exp == max
+ *  - `cast(exp, ty) <= value` ==> if(isnull(exp), null, true)
+ *  - `cast(exp, ty) < value` ==> exp =!= max
+ *
+ * Similarly for the cases when `value == min` and `value < min`.
+ *
+ * Further, the above `if(isnull(exp), null, false)` is represented using conjunction
+ * `and(isnull(exp), null)`, to enable further optimization and filter pushdown to data sources.
+ * Similarly, `if(isnull(exp), null, true)` is represented with `or(isnotnull(exp), null)`.
+ */
+object UnwrapCast extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case l: LogicalPlan => l transformExpressionsUp {
+      case e @ BinaryComparison(_, _) => unwrapCast(e)
+    }
+  }
+
+  private def unwrapCast(exp: Expression): Expression = exp match {
+    case BinaryComparison(Literal(_, _), Cast(_, _, _)) =>
+      // Not a canonical form. In this case we first canonicalize the expression by swapping the
+      // literal and cast side, then process the result and swap the literal and cast again to
+      // restore the original order.
+      def swap(e: Expression): Expression = e match {
+        case GreaterThan(left, right) => LessThan(right, left)
+        case GreaterThanOrEqual(left, right) => LessThanOrEqual(right, left)
+        case EqualTo(left, right) => EqualTo(right, left)
+        case EqualNullSafe(left, right) => EqualNullSafe(right, left)
+        case LessThanOrEqual(left, right) => GreaterThanOrEqual(right, left)
+        case LessThan(left, right) => GreaterThan(right, left)
+        case _ => e
+      }
+
+      swap(unwrapCast(swap(exp)))
+
+    case BinaryComparison(Cast(fromExp, _, _), Literal(value, toType))
+      if canImplicitlyCast(fromExp, toType) =>
+
+      // In case both sides have integral type, optimize the comparison by removing casts or
+      // moving cast to the literal side.
+      simplifyIntegral(exp, fromExp, toType.asInstanceOf[IntegralType], value)
```

Comment:
Is it safe to do the type casting here? We might add more types in `canImplicitlyCast`.

Reply:
If we match `    case BinaryComparison(Cast(fromExp, _, _), Literal(value, toType: IntegralType))` it's safer and no need for casting.

Reply:
Yes currently it's safe since we only allow integral types. Later when we add more types we'll add safe guard here.

We can do the matching on `IntegralType` but I think it might be slightly better to encapsulate the logic in `canImplicitylyCast` so that we can expand it later such as casting from long to double with digits checking etc. But yeah I don't have strong preference on either at this moment.

Reply:
Or we can keep the existing and add the above `toType: IntegralType`, then we can remove the casting.